### PR TITLE
Scroll behaviour adjustments

### PR DESCRIPTION
**Pull request by ASalavei**
What have changed:
- Adjust maximum scroll speed to make animation completed ~30% faster.

### CODE REVIEW

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.uikit.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+
+package androidx.compose.foundation.gestures
+
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.unit.Dp
+
+internal val flingBehaviorVisibilityThresholdDp = Dp(2.0f)
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.uikit.kt:23]: Why common `0.5f` doesn't work? It doesn't seem like a platform thing that requires expect/actual

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.kt [hunk 2]
```diff
@@ -42,4 +43,7 @@ interface FlingBehavior {
      * @return remaining velocity after fling operation has ended
      */
     suspend fun ScrollScope.performFling(initialVelocity: Float): Float
-}
\ No newline at end of file
+}
+
+@Composable
+expect fun flingBehaviorVisibilityThreshold(): Float?
```
Comment (igordmn) [inline compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.kt]: This will crash on Android. We don't build it and reuse the published Google artifacts.

We need to upstream it to AOSP, or make the code internal some way

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/pager/Pager.kt [hunk 3]
```diff
@@ -897,7 +903,7 @@ private class DefaultPagerNestedScrollConnection(
         source: NestedScrollSource
     ): Offset {
         if (source == NestedScrollSource.Fling && available != Offset.Zero) {
-            throw ScrollCancellationException("End of scrollable area reached")
+            throw CancellationException("End of scrollable area reached")
```
Comment (igordmn) [inline compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/pager/Pager.kt]: Let's make a separate PR for that. Otherwise it will be difficult to upstream this PR (and cherry-pick/revert)

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/pager/Pager.kt [hunk 4]
```diff
@@ -466,7 +469,10 @@ object PagerDefaults {
             durationMillis = LowVelocityAnimationDefaultDuration
         ),
         highVelocityAnimationSpec: DecayAnimationSpec<Float> = rememberSplineBasedDecay(),
-        snapAnimationSpec: AnimationSpec<Float> = spring(stiffness = Spring.StiffnessMediumLow),
+        snapAnimationSpec: AnimationSpec<Float> = spring(
+            stiffness = Spring.StiffnessMediumLow,
+            visibilityThreshold = flingBehaviorVisibilityThreshold()
```
Comment (igordmn) [inline compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/pager/Pager.kt:474]: Is it correct to use "fling" concept in the Pager snap physics? (I don't know, so I am asking :))

**Review by igordmn — CHANGES_REQUESTED** [review]
We need to decide what to do with the public function

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.uikit.kt (hunk 1 above)
Comment (ASalavei) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.uikit.kt:23]: The default value for threshold is "null". The value of 0.5 was used in Overscroll animations.
These values are too low and animation curve produces a very long "tail" - basically when animation it running, but without any visual outcome because changes are too trivial. Because of that and with multiple scrolls on a scene it's possible to confuse end user with unexpected behaviour of scrolls. The user may think that scroll action is over, however it still executes in code. For me personally being adopted to iOS timings, it was a huge problem.
Unlike Android, on iOS animation actions stop faster, when the destination goal is reached.
Here, the minimum value of 2.0f (dp/sec) seems a reasonable speed to complete animation because no more significant changes are expected.

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.kt (hunk 2 above)
Comment (ASalavei) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.kt]: Thanks, I made it internal.

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/pager/Pager.kt (hunk 3 above)
Comment (ASalavei) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/pager/Pager.kt]: Sure, reverted

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/pager/Pager.kt (hunk 4 above)
Comment (ASalavei) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/pager/Pager.kt:474]: It seems like yes, the pager animation logic calls `SnapFlingBehavior`

**Comment by ASalavei** [comment]
> We need to decide what to do with the public function

Will make the method 'internal' help here?

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.uikit.kt (hunk 1 above)
Comment (igordmn) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.uikit.kt:23]: Let's write this in the comment ). The explanation is very useful

**Comment by igordmn** [comment]
> Will make the method 'internal' help here?

Yes. I thought that isn't possible as we use it as a default parameter. But it seems it is possible.

**Review by igordmn — APPROVED** [review]

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.uikit.kt (hunk 1 above)
Comment (MatkovIvan) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.uikit.kt:23]: > 2.0f (dp/sec)

It wasn't clear at all

> The value of 0.5 was used in Overscroll animations

It uses 0.5/frame there! Why should we do differently?

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/pager/Pager.kt [hunk 5]
```diff
@@ -345,7 +345,10 @@ object PagerDefaults {
         pagerSnapDistance: PagerSnapDistance = PagerSnapDistance.atMost(1),
         lowVelocityAnimationSpec: AnimationSpec<Float> = LowVelocityAnimationSpec,
         highVelocityAnimationSpec: DecayAnimationSpec<Float> = rememberSplineBasedDecay(),
-        snapAnimationSpec: AnimationSpec<Float> = spring(stiffness = Spring.StiffnessMediumLow),
+        snapAnimationSpec: AnimationSpec<Float> = spring(
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/pager/Pager.kt:348]: It's a change in default parameters of PUBLIC COMMON method. It's considered as a breaking change by `metalava`/Google.

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

**Comment by ASalavei** [comment]
Closing the MR because of upstreaming complications and insignificant improvements of the change.
